File: Erlang/MatrixMultiplication.erl
-module(matmultest).
-export([matmul/2, start/4]).

transpose (Matrix) -> rows(Matrix, []).

rows([], Result) ->
          reverse_rows(Result, []);

rows([Row|OtherRows], Result) ->
                NewResult = makecolumn(Row, Result),
                rows(OtherRows, NewResult).

makecolumn([], Result) -> Result;

makecolumn([H|T], []) ->
          [[H] | makecolumn(T, [])];

makecolumn([H|T], [RH|RT]) ->
          [[H|RH] | makecolumn(T, RT)].


reverse_rows([], Result) -> lists:reverse(Result);

reverse_rows([FirstRow|OtherRows], Result) ->
  reverse_rows(OtherRows, [lists:reverse(FirstRow) | Result]).


matmul(MatrixOne, MatrixTwo) -> multiply(MatrixOne, transpose(MatrixTwo)).

multiply([], _) -> [];

multiply([FirstRow|OtherRows], MatrixTwo) ->
                              [calcRow(FirstRow, MatrixTwo)|multiply(OtherRows, MatrixTwo)].

calcRow(_, []) -> [];

calcRow(Row, [MatrixTwoH|T]) -> [calcCol(Row, MatrixTwoH)|calcRow(Row, T)].

calcCol([],[]) -> 0;

calcCol([RH|RT], [ColH|COlT]) -> RH*ColH + calcCol(RT, COlT).


timeit(F, A) ->
        T1 = erlang:monotonic_time(),
        Val = apply(F, A),
        T2 = erlang:monotonic_time(),
        Time = erlang:convert_time_unit(T2 - T1, native, micro_seconds),
        Time.

start(Fmatrix1, Fmatrix2, Counter, Result) ->
   if Counter > 0 ->
           {ok, BinaryList1} = file:consult(Fmatrix1),
           {ok, BinaryList2} = file:consult(Fmatrix2),
           Matrix1 = hd(BinaryList1),
           Matrix2 = hd(BinaryList2),
           NewResult = timeit(fun matmul/2, [Matrix1, Matrix2]),
           start(Fmatrix1, Fmatrix2, Counter-1, [NewResult|Result]);
      true -> Result
   end.

File: Haskell/MatrixMultiplication.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE BangPatterns #-}
import Control.Exception
import Formatting
import Formatting.Clock
import System.Clock
import Control.DeepSeq
import System.IO

transpose :: [[a]] -> [[a]]
transpose m = rows m []

rows :: [[a]] -> [[a]] -> [[a]]
rows [] result = reverserows result []
rows (row:otherRows) result = let newresult = makecolumn row result in rows otherRows newresult

makecolumn :: [t] -> [[t]] -> [[t]]
makecolumn [] result = result
makecolumn (row:otherRows) [] = [row]:makecolumn otherRows []
makecolumn (row:otherRows) (fResultRow:otherResultRows) = (row:fResultRow): makecolumn otherRows otherResultRows

reverserows :: [[a]] -> [[a]] -> [[a]]
reverserows [] result = reverse result
reverserows (row:otherRows) result = reverserows otherRows (reverse row : result)

matmul :: Num a => [[a]] -> [[a]] -> [[a]]
matmul matrixOne matrixTwo = multiply matrixOne (transpose matrixTwo)

multiply :: Num a => [[a]] -> [[a]] -> [[a]]
multiply [] x = []
multiply (firstRow: otherRows) matrixTwo = calcRow firstRow matrixTwo : multiply otherRows matrixTwo

calcRow :: Num a => [a] -> [[a]] -> [a]
calcRow x [] = []
calcRow row (matrixTwoRow:otherRows) = calcCol row matrixTwoRow : calcRow row otherRows

calcCol :: Num a => [a] -> [a] -> a
calcCol [] [] = 0
calcCol (row: otherRows) (matrixTwoRow:matrixTwootherRows) = row*matrixTwoRow + calcCol otherRows matrixTwootherRows

repeater matrixOne matrixTwo 0 result = return (result)

repeater matrixOne matrixTwo counter result = do
                                      start <- getTime Monotonic
                                      let !e = matmul matrixOne matrixTwo
                                      print(last e)
                                      end <- getTime Monotonic
                                      let newTime = toNanoSecs (diffTimeSpec start end)
                                      repeater matrixOne matrixTwo (counter-1) (newTime:result)
